File: Aeris.cpp
#include <iostream>
#include <cstdlib>
#include <cpr/cpr.h>
#include <nlohmann/json.hpp>
#include <fstream>
#include <algorithm>
using namespace std;
using json = nlohmann::json;

int main(){
    //string API_KEY = getenv("API_KEY");
    ifstream file("test.txt");
    string API_KEY;
    getline(file, API_KEY);
    string link; 
    string city;
    string state;
    string country;
    string unit;


    cout << "Enter a city: ";                                               // Format the city to be link friendly, as cpr seems to not automatically handle spaces
    getline(cin, city);
    for (int i = 0; i < city.length(); i++){
        if (city[i] == ' '){
            city[i] = '+';
        }
    }

    cout << "Enter a state code (only for US, type -1 to disregard): ";     // Enter the state code (NY, CA, etc.)
    getline(cin, state);

    cout << "Enter a country code: ";
    getline(cin, country);                                                  // Enter the country code (CN, QA, MX, etc.)

    cout << "Metric or imperial?: ";                                        // User must input metric or imperial
    getline(cin, unit);
    transform(unit.begin(), unit.end(), unit.begin(), ::tolower);

    if (unit == "metric" || unit == "imperial"){
        if (state == "-1"){
            link = "http://api.openweathermap.org/data/2.5/weather?q=" + city + "," + country + "&units=" + unit + "&lang=en&appid=" + API_KEY;
        }
        else{
            link = "http://api.openweathermap.org/data/2.5/weather?q=" + city + "," + state + "," + country + "&units=" + unit + "&lang=en&appid=" + API_KEY;
        }
    }
    else{
        cout << "Invalid unit!" << endl;
    }

    cpr::Response r = cpr::Get(cpr::Url{link});
    json response = json::parse(r.text);
    cout << setw(4) << "Temperature: " << response["main"]["temp"]  << endl << setw(4) << "Weather: " << response["weather"][0]["main"] << endl;
}

File: Aeris.py
from flask import Flask, request, render_template, jsonify
from dotenv import load_dotenv
import json, requests, os

load_dotenv()
api_key = os.getenv("API_KEY")
location_api_key = os.getenv("LOCATION_API_KEY")
mapbox_api = os.getenv("MAPBOX_API")
app = Flask(__name__)

@app.route("/about", methods=["GET"])
def about():
    return render_template("about.html")

@app.route("/search", methods=["GET"])
def search():
    query = request.args.get('q')
    
    if not query:
        return '[]'
    
    geourl = f"https://api.mapbox.com/search/geocode/v6/forward?q={query}&access_token={mapbox_api}&language=en&limit=20&types=place"
    response = requests.request("GET", geourl)
    
    return response.text

@app.route("/weather", methods=["GET"])
def weather():
    city = request.args.get("city")
    state = request.args.get("state")
    country = request.args.get("country")
    unit = request.args.get("unit")
    
    if state and state.strip():
        url = f"http://api.openweathermap.org/data/2.5/weather?q={city},{state},{country}&units={unit}&lang=en&appid={api_key}"
    else:
        url = f"http://api.openweathermap.org/data/2.5/weather?q={city},{country}&units={unit}&lang=en&appid={api_key}"
    
    response = json.loads(requests.request("GET", url).text)
    temperature = response["main"]["temp"]
    desc = response["weather"][0]["description"]
    icon = response["weather"][0]["icon"]

        
    if (unit.lower() == "imperial"):
        tempChar = "°F"
    elif (unit.lower() == "metric"):
        tempChar = "°C"
    else:
        tempChar = ""

    return {"temperature": temperature, "desc": desc, "tempChar": tempChar, "icon": icon}


@app.route("/", methods=["GET", "POST"])
def aeris():
    if request.method == "POST":
        city = request.form.get("city")
        state = request.form.get("state")
        country = request.form.get("country")
        unit = request.form.get("unit")
        
        if state and state.strip():
            url = f"http://api.openweathermap.org/data/2.5/weather?q={city},{state},{country}&units={unit}&lang=en&appid={api_key}"
        else:
            url = f"http://api.openweathermap.org/data/2.5/weather?q={city},{country}&units={unit}&lang=en&appid={api_key}"
        
        response = json.loads(requests.request("GET", url).text)

        temperature = response["main"]["temp"]
        desc = response["weather"][0]["description"]
        icon = response["weather"][0]["icon"]
        
        if (unit.lower() == "imperial"):
            tempChar = "°F"
        elif (unit.lower() == "metric"):
            tempChar = "°C"
        else:
            tempChar = ""
            
        return render_template("Aeris.html", city=city, state=state, country=country, tempChar=tempChar, 
                                 unit=unit, temperature=temperature, desc=desc, icon=icon)
    
    return render_template("Aeris.html")

app.run(host="0.0.0.0", port = 5000)